src/devices/Mpu9250/Mpu6500.cs

```diff
@@ -0,0 +1,899 @@
+        /// Report passing Self-Test if the following criteria fulfilled. 
+        /// Axis   | Pass criteria 
+        /// X-gyro | |GXOFFSET| ≤ 20dps
+        /// Y-gyro | |GYOFFSET| ≤ 20dps
+        /// Z-gyro | |GZOFFSET| ≤ 20dps
+        /// ]]>
+        /// </summary>
+        /// <returns>the gyroscope and accelerometer vectors</returns>
+        public (Vector3 gyroscopeAverage, Vector3 accelerometerAverage) RunGyroscopeAccelerometerSelfTest()
+        {
+            // Used for the number of cycles to run the test
+            // Value is 200 according to documentation AN-MPU-9250A-03
+            const int numCycles = 200;
+
+            Vector3 accAverage = new Vector3();
+            Vector3 gyroAvegage = new Vector3();
+            Vector3 accSelfTestAverage = new Vector3();
+            Vector3 gyroSelfTestAverage = new Vector3();
+            Vector3 gyroSelfTest = new Vector3();
+            Vector3 accSelfTest = new Vector3();
+            Vector3 gyroFactoryTrim = new Vector3();
+            Vector3 accFactoryTrim = new Vector3();
+            // Tests done with slower GyroScale and Accelerator so 2G so value is 0 in both cases
+            byte gyroAccScale = 0;
+
+            // Setup the registers for Gyroscope as in documentation 
+            // DLPF Config | LPF BW | Sampling Rate | Filter Delay 
+            // 2           | 92Hz   | 1kHz          | 3.9ms 
+            WriteRegister(Register.SMPLRT_DIV, 0x00);
+            WriteRegister(Register.CONFIG, 0x02);
+            GyroscopeRange = GyroscopeRange.Range0250Dps;
+            // Set full scale range for the gyro to 250 dps
+            // Setup the registers for accelerometer as in documentation
+            // DLPF Config | LPF BW | Sampling Rate | Filter Delay 
+            // 2           | 92Hz   | 1kHz          | 7.8ms 
+            WriteRegister(Register.ACCEL_CONFIG_2, 0x02);
+            AccelerometerRange = AccelerometerRange.Range02G;
+
+            // Read the data 200 times as per the documentation page 5
+
+            for (int reading = 0; reading < numCycles; reading++)
+            {
+                gyroAvegage = GetRawGyroscope();
+                accAverage = GetRawAccelerometer();
+            }
+
+            accAverage /= numCycles;
+            gyroAvegage /= numCycles;
+
+            // Set USR_Reg: (1Bh) Gyro_Config, gdrive_axisCTST [0-2] to b111 to enable Self-Test. 
+            WriteRegister(Register.ACCEL_CONFIG, 0xE0);
+            // Set USR_Reg: (1Ch) Accel_Config, AX/Y/Z_ST_EN   [0-2] to b111 to enable Self-Test. 
+            WriteRegister(Register.GYRO_CONFIG, 0xE0);
+            // Wait 20ms for oscillations to stabilize 
+            Thread.Sleep(20);
+
+            // Read the gyroscope and accelerometer output at a 1kHz rate and average 200 readings. 
+            // The averaged values will be the LSB of GX_ST_OS, GY_ST_OS, GZ_ST_OS, AX_ST_OS, AY_ST_OS and AZ_ST_OS in the software
```

Comment:
similar as above - possibly some of the code might be possible to re-use

Reply:
Except running something similar, I don't think there is a need to reuse. But I have hard time to see what can be done in a similar way

Reply:
I don't feel strongly about this comment